File: 2021/ts/three/main.ts
import { solve, test } from "../lib/utils.ts";

async function run() {
  const day = "three";
  await test(day, partOne, 198);
  await solve(day, partOne, 1131506);
  await test(day, partTwo, 230);
  await solve(day, partTwo, 7863147);
}

function countBitsAtPosition(
  input: Array<Array<number>>,
  bitPosition: number
): Array<number> {
  let zeros = 0;
  let ones = 0;
  input.forEach((val) => {
    if (val[bitPosition] === 0) {
      zeros++;
    } else {
      ones++;
    }
  });
  return [zeros, ones];
}

function partOne(inputRaw: Array<string>): number {
  const input = inputRaw.map((val) => val.split("").map((x) => parseInt(x, 2)));
  const gammaBits: Array<number> = [];
  const epsilonBits: Array<number> = [];
  const bitPositions = input[0].length;
  for (let i = 0; i < bitPositions; i++) {
    const [zeros, ones] = countBitsAtPosition(input, i);
    if (ones >= zeros) {
      gammaBits.push(1);
      epsilonBits.push(0);
    } else {
      gammaBits.push(0);
      epsilonBits.push(1);
    }
  }

  const gammaRate = parseInt(gammaBits.join(""), 2);
  const epsilonRate = parseInt(epsilonBits.join(""), 2);
  return gammaRate * epsilonRate;
}

function getRating(
  input: Array<Array<number>>,
  bitPrecedence: Array<number>
): number {
  let ratings = [...input];
  let i = 0;
  while (ratings.length > 1) {
    const [zeros, ones] = countBitsAtPosition(ratings, i);
    const candidateRatings: Array<Array<number>> = [];
    ratings.forEach((val) => {
      const bit = val[i];
      if (ones >= zeros) {
        if (bit === bitPrecedence[0]) {
          candidateRatings.push(val);
        }
      } else if (ones < zeros) {
        if (bit === bitPrecedence[1]) {
          candidateRatings.push(val);
        }
      }
    });
    ratings = candidateRatings;
    i++;
  }
  return parseInt(ratings.flat().join(""), 2);
}

function partTwo(inputRaw: Array<string>): number {
  const input = inputRaw.map((val) => val.split("").map((x) => parseInt(x, 2)));

  const o2 = getRating(input, [1, 0]);
  const co2 = getRating(input, [0, 1]);
  return o2 * co2;
}

await run();

File: 2021/rust/three/src/main.rs
use std::time::Instant;

fn solve_part(solver: &dyn Fn(std::str::Lines) -> i32, expected: i32) {
    let lines = include_str!("../../../input/three.txt").lines();
    run_part(lines, solver, expected)
}

fn test_part(solver: &dyn Fn(std::str::Lines) -> i32, expected: i32) {
    let lines = include_str!("../../../input/test/three.txt").lines();
    run_part(lines, solver, expected)
}

fn run_part(lines: std::str::Lines, solver: &dyn Fn(std::str::Lines) -> i32, expected: i32) {
    let start = Instant::now();
    let answer = solver(lines.clone());
    let duration = start.elapsed();
    println!(
        "Part one answer: {}, expected {}. Completed in: {:?}",
        answer, expected, duration
    );
    assert_eq!(answer, expected);
}

fn main() {
    test_part(&part_one, 198);
    solve_part(&part_one, 1131506);

    test_part(&part_two, 230);
    solve_part(&part_two, 7863147);
}

fn get_int_from_binary_vec(bits: &Vec<u8>) -> i32 {
    isize::from_str_radix(&bits.iter().map(|x| x.to_string()).collect::<String>(), 2).unwrap()
        as i32
}

fn count_bits_at_position(bit_lines: &Vec<Vec<u8>>, bit_position: usize) -> (i32, i32) {
    let mut ones = 0;
    let mut zeros = 0;
    for bits in bit_lines {
        if bits[bit_position] == 1 {
            ones = ones + 1;
        } else {
            zeros = zeros + 1;
        }
    }
    (zeros, ones)
}

fn get_rating(bit_lines: &Vec<Vec<u8>>, bit_precedence: (u8, u8)) -> i32 {
    let mut ratings = bit_lines.clone();
    let mut i = 0;

    while ratings.len() > 1 {
        let (zeros, ones) = count_bits_at_position(&ratings, i);
        let mut candidate_ratings: Vec<Vec<u8>> = vec![];
        for rating in ratings {
            let bit = rating[i];
            if ones >= zeros {
                if bit == bit_precedence.0 {
                    candidate_ratings.push(rating);
                }
            } else {
                if bit == bit_precedence.1 {
                    candidate_ratings.push(rating);
                }
            }
        }
        ratings = candidate_ratings;
        i = i + 1;
    }
    get_int_from_binary_vec(ratings.iter().next().unwrap())
}

fn part_one(lines: std::str::Lines) -> i32 {
    let bit_lines = lines
        .map(|x| {
            x.chars()
                .map(|c| c.to_digit(2).unwrap() as u8)
                .collect::<Vec<u8>>()
        })
        .collect::<Vec<_>>();
    let mut gamma_bits: Vec<u8> = vec![];
    let mut epsilon_bits: Vec<u8> = vec![];
    let bit_positions = bit_lines[0].len();
    let mut i = 0;
    while i < bit_positions {
        let (zeros, ones) = count_bits_at_position(&bit_lines, i);
        if ones >= zeros {
            gamma_bits.push(1);
            epsilon_bits.push(0);
        } else {
            gamma_bits.push(0);
            epsilon_bits.push(1);
        }
        i = i + 1;
    }
    let gamma_rate = get_int_from_binary_vec(&gamma_bits);
    let epsilon_rate = get_int_from_binary_vec(&epsilon_bits);
    epsilon_rate * gamma_rate
}

fn part_two(lines: std::str::Lines) -> i32 {
    let bit_lines = lines
        .map(|x| {
            x.chars()
                .map(|c| c.to_digit(2).unwrap() as u8)
                .collect::<Vec<u8>>()
        })
        .collect::<Vec<_>>();
    let o2_rating = get_rating(&bit_lines, (1, 0));
    let co2_rating = get_rating(&bit_lines, (0, 1));
    o2_rating * co2_rating
}
